Update rule: avoid_noarch

msentissi opened the pull request:
This PR updates the `avoid_noarch` LintCheck:
- For the root and each output, `/build/noarch` is checked and a warning is produced if its value is `python`.
  - Rebuilds and exceptions are skipped.
- In each case, the auto-fix:
  - removes `noarch: python`
  - attempts to extract the `python` version from the run dependencies
  - adds `python` as a `host`, `run`, and `test` requirement
  - updates the skip statement if a python version was extracted successfully
- Standard and auto-fix test cases are added.

schuylermartin45 commented on anaconda_linter/lint/check_build_help.py:
```diff
@@ -619,83 +622,129 @@ class avoid_noarch(LintCheck):
+        if not data:
+            return False
+        package_path, name = data
+        recipe = self.unrendered_recipe
+
+        # Remove build/noarch
+        removed_noarch = recipe.patch(
```
These can be made final

msentissi commented:
@schuylermartin45 I addressed your comments. Please note that I also disabled `stdlib_must_be_in_build` tests temporarily since `Percy` variants rendering has been updated in the most recent version in a way that requires that check to be adapted. I will work on it next and re-enable full testing.

schuylermartin45 commented on anaconda_linter/lint/check_build_help.py:
```diff
@@ -619,83 +622,129 @@ class avoid_noarch(LintCheck):
+                continue
+            if name in NOARCH_ALLOWLIST:
+                continue
+            try:
+                build_number: Final[str] = recipe.get_value(recipe.append_to_path(package_path, "/build/number"))
+            except KeyError:
+                try:
+                    build_number: Final[str] = recipe.get_value("/build/number")
+                except KeyError:
+                    self.message(title_in="Failed to determine build number, cannot run this check.")
+                    continue
```
I guess I was talking about making a function that both tests could call. I don't think we want to make tests and checks dependent on each other entirely. I think at some point we will want to parallelize these linter tests, like we do in CRM. Assuming we ever get to the point where the tests feel slow.

As for the `Final` discussion: I think the anaylzer just isn't running in our pre-commit/commit checks in this project. I don't think this project has the same level of code checking as CRM. I was only able to back-port so much when I last touched this project, and idk what state CRM was in at the time. It could very well be CRM has much newer and stricter rules for code formatting/linting/mypy.

I believe prior to my involvement this project _only_ had tests as checks.

I'm going to go ahead and ignore this point for approval as if I'm right, the static analyzer will let you know when it is running.
